crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -298,6 +428,64 @@ impl<'stmt> CFGBuilder<'stmt> {
         self.cfg.blocks.push(BlockData::default())
     }
 
+    /// Returns index of block where control flow should proceed
+    /// at the current depth.
+    ///
+    /// Creates a new block if there are remaining statements, otherwise
+    /// returns provided default.
+    fn next_or_default_block(
+        &mut self,
+        remaining_stmts: &'stmt [Stmt],
+        default: BlockId,
+    ) -> BlockId {
+        if remaining_stmts.is_empty() {
+            default
+        } else {
+            self.new_block()
+        }
+    }
+
+    /// Creates a new block to handle entering and exiting a loop body.
+    fn new_loop_guard(&mut self) -> BlockId {
+        let Some(currblock) = self.cfg.blocks.get_mut(self.current) else {
+            return self.cfg.blocks.push(BlockData {
+                kind: BlockKind::LoopGuard,
+                ..BlockData::default()
+            });
+        };
+        if matches!(currblock.kind, BlockKind::Generic) && currblock.stmts.is_empty() {
+            currblock.kind = BlockKind::LoopGuard;
+            self.current
+        } else {
+            self.cfg.blocks.push(BlockData {
+                kind: BlockKind::LoopGuard,
+                ..BlockData::default()
+            })
+        }
+    }
+
+    /// Returns the current loop exit block without removing it.
+    fn loop_exit(&self) -> Option<BlockId> {
+        self.loops.last().map(|ctxt| ctxt.exit)
+    }
+    /// Returns the current loop guard block without removing it.
+    fn loop_guard(&self) -> Option<BlockId> {
+        self.loops.last().map(|ctxt| ctxt.guard)
+    }
+
+    /// Pushes a block onto the loop exit stack.
+    /// This block represents where control should flow when encountering a
+    /// 'break' statement within a loop.
+    fn push_loop(&mut self, guard: BlockId, exit: BlockId) {
+        self.loops.push(LoopContext { guard, exit });
+    }
+
+    /// Pops and returns the most recently pushed loop exit block.
+    /// This is called when finishing the processing of a loop construct.
+    fn pop_loop(&mut self) -> Option<LoopContext> {
+        self.loops.pop()
```

Comment:
Nit: I find it useful to add assertion that "stack" variables are empty once the builder phase is completed. It helped me in the past to find bugs. You could add a `asser_eq!(self.loops, &[])` to `CfgBuilder::finish`